Update/react native60

makedirectory opened the pull request:
Happy to leave comments if necessary on what was changed/updated

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/react-native-firebase-starter?pullRequest=135) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/react-native-firebase-starter?pullRequest=135) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/react-native-firebase-starter?pullRequest=135) it.</sub>

makedirectory commented:
iOS testers may need to manually modify the project configurations here if receiving an error related to RCTLog not found: 

<img width="877" alt="Screen Shot 2019-09-08 at 4 33 57 PM" src="https://user-images.githubusercontent.com/30823997/64494097-9a582880-d256-11e9-8cc4-e1a12c3b3872.png">

RNFirebaseStarterTests for both debug and release should be as follows

<img width="867" alt="Screen Shot 2019-09-08 at 4 36 07 PM" src="https://user-images.githubusercontent.com/30823997/64494114-c673a980-d256-11e9-9801-0f822ecff8aa.png">

mikehardy commented:
This is fantastic, and sorely needed, thank you! I won't have time to review it for a couple days while I handle a react-native-device-info major release, but if I test it locally and it works I'll be +1 pretty easy

One question about your follow on hint that iOS testers might need to adjust things - that wasn't the case with the rn59 version, is there something inherent in rn60 that makes manual adjustment necessary? I might be missing a subtlety but I don't know why that would be needed except maybe for setting development teams...

makedirectory commented:
> One question about your follow on hint that iOS testers might need to adjust things - that wasn't the case with the rn59 version, is there something inherent in rn60 that makes manual adjustment necessary? I might be missing a subtlety but I don't know why that would be needed except maybe for setting development teams...

I believe this has something to do with the use of Pods for linking React dependencies. Every time I do a `pod install` the config for the Test project is reset too 'None'. So that's why I've included the comment above. This may be a new issue with RN due to the switch but I'm not familiar enough to file an issue, if the consensus determines this is an issue I'm happy to do so.

A few other things to note: 

(1) I could not get Android version to build without adjusting org.gradle.jvmargs within gradle.properties. I set it to 4096m but this may be able to be set lower.

(2) For Android you will need to include the Google Ads ID or the build will crash without loading the bundle. Little hard to trace this, had to use `adb logcat` to view the device logs.

(3) On Android, the RNFirebase repo is no longer linked in /app/build.gradle or in the MainApplication.java but I didn't remove them, just comment out.

(4) I didn't add a react-native-config.js file to the root of the application because there is no rnpm configuration in place. If you want this added, please let me know.

(5) On iOS, I didn't remove the specific versions for the Firebase Pods but I think that can be done without any issues followed by a `pod update` - happy to add this to the PR.

Apologies for making one large commit, I should have broken it down into multiple. If I can think of anything else I will be sure to leave an additional comment.

mikehardy commented on android/gradle.properties:
```diff
@@ -10,9 +10,12 @@
 # The setting is particularly useful for tweaking memory settings.
 # Default value: -Xmx10248m -XX:MaxPermSize=256m
-# org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
+org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=4096m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
```
I'm not sure it's valid to have maxpermsize higher than mx? I'm building successfully with this `org.gradle.jvmargs=-Xmx2048M -Dkotlin.daemon.jvm.options\="-Xmx3072M" -XX\:MaxPermSize\=768m -XX\:+HeapDumpOnOutOfMemoryError -Dfile.encoding\=UTF-8` so maybe just going to 768 is okay

makedirectory replied:
Updated to `-XX:MaxPermSize=768m`

mikehardy commented on ios/Podfile:
```diff
@@ -1,12 +1,38 @@
   # Required by RNFirebase
+  pod 'RNFirebase', path: '../node_modules/react-native-firebase/ios'
+
   pod 'Firebase/Core', '~> 6.3.0'
```
might as well move these to 6.8.0 now and at least start current :-)

makedirectory replied:
I did try and remove this, cleared all derived data and build folder. Was not able to build without - screenshot below:

<img width="434" alt="Screen Shot 2019-09-09 at 9 37 39 PM" src="https://user-images.githubusercontent.com/30823997/64578471-83483200-d34d-11e9-8928-e11b6da259a3.png">

mikehardy commented on package.json:
```diff
@@ -9,21 +9,26 @@
     "babel-jest": "^24.8.0",
+    "eslint": "^6.1.0",
     "fs-extra": "^7.0.1",
     "jest": "^24.8.0",
```
this and babel-jest at 24.8.0 are specifically problematic and are best at 24.9.0 now (I think yarn handles the range well but npm does not and you get stuck on the old version which has terrible performance) with node v12.4-12.8 among other problems

makedirectory replied:
Updated both to 24.9.0 - also I always use yarn!

mikehardy commented on ios/Podfile:
```diff
@@ -1,12 +1,38 @@
+  pod 'Folly', :podspec => '../node_modules/react-native/third-party-podspecs/Folly.podspec'
+
   # Required by RNFirebase
+  pod 'RNFirebase', path: '../node_modules/react-native-firebase/ios'
```
This is actually not required in auto-linking. I no longer have this line, it autolinks in. But the rest of them are important (any `Firebase/xxx` module you need)

makedirectory replied:
I did try and remove this, cleared all derived data and build folder. Was not able to build without - screenshot below:

<img width="434" alt="Screen Shot 2019-09-09 at 9 37 39 PM" src="https://user-images.githubusercontent.com/30823997/64578471-83483200-d34d-11e9-8928-e11b6da259a3.png">

mikehardy replied:
This is really the only remaining thing I'm hung up on. I definitely don't have this and it definitely works. I wonder if the project is still linked and needs to be unlinked in order for the auto-linking to take effect during pod install? That might explain the RCTLog thing too (which is odd, to me)

Wait! I think I have it! I believe it's missing use_native before the end of the Podfile https://github.com/react-native-community/rn-diff-purge/blob/release/0.60.5/RnDiffApp/ios/Podfile#L35 so it's not linking. I bet if you add that, you can take out the core module (it should auto-link) and it might fix RCTLog too?

makedirectory replied:
You are correct. Both issues resolved! Was missing the RNFirebaseStarterTests target in the podfile

makedirectory replied:
I also removed the reference to this issue in the readme file as it's no longer needed.

mikehardy commented on android/build.gradle:
```diff
@@ -16,23 +16,27 @@ buildscript {
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.4.1'
+        classpath('com.android.tools.build:gradle:3.4.1')
```
I'd bump all these up - gradle 3.5.0, google-services 4.3.2, etc

makedirectory replied:
Update gradle and google services! perf-plugin and fabric were both updated in the initial commit.

mikehardy commented on android/app/build.gradle:
```diff
@@ -154,45 +202,45 @@ dependencies {
     //noinspection GradleCompatible
-    implementation "com.google.firebase:firebase-core:16.0.9"
-    implementation "com.google.android.gms:play-services-base:16.1.0"
+    implementation "com.google.firebase:firebase-core:17.2.0"
```
strongly recommend going to the 'bom' style for these (and any projects you are using firebase on) like so (this is mine, I may not have all the modules desired, but the style is illustrative:

```groovy
    implementation platform("com.google.firebase:firebase-bom:22.1.0")
    implementation "com.google.firebase:firebase-core"
    implementation "com.google.firebase:firebase-auth"
    implementation "com.google.firebase:firebase-analytics"
    implementation "com.google.firebase:firebase-perf"
    implementation "com.google.firebase:firebase-config"
    implementation "com.google.firebase:firebase-functions"
    implementation "com.google.firebase:firebase-storage"
    implementation "com.google.firebase:firebase-dynamic-links"
    implementation "com.google.firebase:firebase-firestore"
    implementation "com.google.firebase:firebase-messaging"
    implementation 'com.crashlytics.sdk.android:crashlytics'


```

makedirectory replied:
I think this should be right, might want to double check. Added

implementation platform("com.google.firebase:firebase-bom:22.2.0")

And removed all versions from other dependencies.

mikehardy commented:
> > One question about your follow on hint that iOS testers might need to adjust things - that wasn't the case with the rn59 version, is there something inherent in rn60 that makes manual adjustment necessary? I might be missing a subtlety but I don't know why that would be needed except maybe for setting development teams...
> 
> I believe this has something to do with the use of Pods for linking React dependencies. Every time I do a `pod install` the config for the Test project is reset too 'None'. So that's why I've included the comment above. This may be a new issue with RN due to the switch but I'm not familiar enough to file an issue, if the consensus determines this is an issue I'm happy to do so.

Interesting - as long as it is documented very clearly in the README I don't mind

> 
> A few other things to note:
> 
> (1) I could not get Android version to build without adjusting org.gradle.jvmargs within gradle.properties. I set it to 4096m but this may be able to be set lower.

I commented on that I think 768 is alright

> 
> (2) For Android you will need to include the Google Ads ID or the build will crash without loading the bundle. Little hard to trace this, had to use `adb logcat` to view the device logs.

If you look at https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh I just put the official sample admob ids in there and it works great. That one is a snag that catches people out, yes. Even harder to figure out on iOS where logging is harder to obtain 

> 
> (3) On Android, the RNFirebase repo is no longer linked in /app/build.gradle or in the MainApplication.java but I didn't remove them, just comment out.

I'd just remove them, them but you still have to have it in settings.gradle so you can pull in the objects in MainApplication for the modules, plus to implement messaging or whatever

> (4) I didn't add a react-native-config.js file to the root of the application because there is no rnpm configuration in place. If you want this added, please let me know.

Nah - I think that's just for overrides? I don't have one, personally - it's not needed for any of the modules I use at this point, including react-native-firebase

> 
> (5) On iOS, I didn't remove the specific versions for the Firebase Pods but I think that can be done without any issues followed by a `pod update` - happy to add this to the PR.

I think having specific versions is still useful but I'd make them all current as I commented

> 
> Apologies for making one large commit, I should have broken it down into multiple. If I can think of anything else I will be sure to leave an additional comment.

I'm not bothered :-), just really happy you took the time to do it, rn59->rn60 isn't the funnest conversion but people really use this starter heavily

makedirectory commented:
I updated the readme.md file to reference the comment above regarding the test configuration for anybody that has issues. It does reset the configuration every single time a `pod install` is performed. 

Only use the react-native-config.js file to add custom fonts, nothing else at this point and it's not needed for this project. 

Do you think it would make sense to add the sample config for the firebase-ads package? 
IE: add 
`<meta-data
            android:name="com.google.android.gms.ads.APPLICATION_ID"
            android:value="ca-app-pub-3940256099942544~3347511713"/>`
to the AndroidManifest.xml file... 

Also worth noting that I added in a Run Script to the iOS build to start metro builder, will make a comment in the code for reference.

makedirectory commented on ios/RNFirebaseStarter.xcodeproj/project.pbxproj:
```diff
@@ -1186,6 +402,23 @@
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "if nc -w 5 -z localhost 8081 ; then\nif ! curl -s \"http://localhost:8081/status\" | grep -q \"packager-status:running\" ; then\necho \"Port 8081 already in use, packager is either not running or not running correctly\"\nexit 2\n\nfi\nelse\nopen \"$SRCROOT/../node_modules/react-native/scripts/launchPackager.command\" || echo \"Can't start packager automatically\"\nfi\n\n";
```
@mikehardy Run Script to start metro builder at the end of the build process

<img width="1147" alt="Screen Shot 2019-09-09 at 10 13 37 PM" src="https://user-images.githubusercontent.com/30823997/64578960-1897f600-d34f-11e9-837c-f337259102a2.png">

Script below for convenience: 

if nc -w 5 -z localhost 8081 ; then
if ! curl -s "http://localhost:8081/status" | grep -q "packager-status:running" ; then
echo "Port 8081 already in use, packager is either not running or not running correctly"
exit 2

fi
else
open "$SRCROOT/../node_modules/react-native/scripts/launchPackager.command" || echo "Can't start packager automatically"
fi

mikehardy commented:
> Do you think it would make sense to add the sample config for the firebase-ads package?
> IE: add
> `<meta-data android:name="com.google.android.gms.ads.APPLICATION_ID" android:value="ca-app-pub-3940256099942544~3347511713"/>`
> to the AndroidManifest.xml file...

Definitely - and in the Info.plist too - the goal of the starter is you follow a couple easy steps and the thing runs. A crash followed by doc-hunting will mean issues open in the repo - no fun. If you count on ad revenue, you'll figure it out soon enough to put in your ID

mikehardy commented on android/app/build.gradle:
```diff
@@ -153,46 +201,46 @@ dependencies {
-    implementation "com.google.firebase:firebase-core:16.0.9"
-    implementation "com.google.android.gms:play-services-base:16.1.0"
+    implementation "com.google.firebase:firebase-core"
+    implementation "com.google.android.gms:play-services-base:17.1.0"
```
I just double-checked my work project, I no longer have this in there. It must be a transitive dependency, because I build all day long and it works without it. Could you try removing it? That would be the last remaining non-BOM versioned thing, which is a real simplifying factor for people I think

makedirectory replied:
Remove `implementation "com.google.android.gms:play-services-base:17.1.0"` or just the version number appended to the end?

mikehardy replied:
The whole thing - I don't have that dependency at all in my android/app/build.gradle any more

mikehardy replied:
:crossed_fingers:

makedirectory replied:
Tested and working without. Pushed change!

makedirectory commented:
> > Do you think it would make sense to add the sample config for the firebase-ads package?
> > IE: add
> > `<meta-data android:name="com.google.android.gms.ads.APPLICATION_ID" android:value="ca-app-pub-3940256099942544~3347511713"/>`
> > to the AndroidManifest.xml file...
> 
> Definitely - and in the Info.plist too - the goal of the starter is you follow a couple easy steps and the thing runs. A crash followed by doc-hunting will mean issues open in the repo - no fun. If you count on ad revenue, you'll figure it out soon enough to put in your ID

Added in https://github.com/invertase/react-native-firebase-starter/pull/135/commits/a836c29a40dc92c3823a07b7ce2af608f7cf592e and updated the readme

mikehardy commented:
I love this - @Salakar + @Ehesp I am human and miss things, but I've scanned this well and @makedirectory has done what I think is some great work here. I'd merge it pending a second set of eyeballs
